crates/astria-sequencer/src/app.rs

```diff
@@ -0,0 +1,195 @@
+
+        stateless.await??;
+        stateful.await??;
+
+        // At this point, the stateful checks should have completed,
+        // leaving us with exclusive access to the Arc<State>.
+        let mut state_tx = self
+            .state
+            .try_begin_transaction()
+            .expect("state Arc should be present and unique");
+
+        tx.execute(&mut state_tx).await?;
+        state_tx.apply();
+
+        let height = self
+            .state
+            .get_block_height()
+            .await
+            .expect("block height should be set");
+        info!(?tx, ?height, "executed transaction");
+        Ok(vec![])
+    }
+
+    pub async fn end_block(&mut self, _end_block: &abci::request::EndBlock) -> Vec<abci::Event> {
+        let state_tx = StateDelta::new(self.state.clone());
+        let mut arc_state_tx = Arc::new(state_tx);
+
+        // call end_block on all components
+        AccountsComponent::end_block(&mut arc_state_tx, _end_block).await;
+        let state_tx = Arc::try_unwrap(arc_state_tx)
+            .expect("components should not retain copies of shared state");
+        self.apply(state_tx)
+    }
+
+    pub async fn commit(&mut self, storage: Storage) -> AppHash {
+        // We need to extract the State we've built up to commit it.  Fill in a dummy state.
+        let dummy_state = StateDelta::new(storage.latest_snapshot());
+        let state = Arc::try_unwrap(std::mem::replace(&mut self.state, Arc::new(dummy_state)))
+            .expect("we have exclusive ownership of the State at commit()");
+
+        // Commit the pending writes, clearing the state.
+        let jmt_root = storage
+            .commit(state)
+            .await
+            .expect("must be able to successfully commit to storage");
+
+        let app_hash: AppHash = jmt_root;
+        tracing::debug!(?app_hash, "finished committing state");
+
+        // Get the latest version of the state, now that we've committed it.
+        self.state = Arc::new(StateDelta::new(storage.latest_snapshot()));
+
+        app_hash
+    }
+
+    // StateDelta::apply only works when the StateDelta wraps an underlying
+    // StateWrite.  But if we want to share the StateDelta with spawned tasks,
+    // we usually can't wrap a StateWrite instance, which requires exclusive
+    // access. This method "externally" applies the state delta to the
+    // inter-block state.
+    //
+    // Invariant: state_tx and self.state are the only two references to the
+    // inter-block state.
+    fn apply(&mut self, state_tx: StateDelta<InterBlockState>) -> Vec<Event> {
+        let (state2, mut cache) = state_tx.flatten();
+        std::mem::drop(state2);
+        // Now there is only one reference to the inter-block state: self.state
+
+        let events = cache.take_events();
+        cache.apply_to(
+            Arc::get_mut(&mut self.state).expect("no other references to inter-block state"),
```

Comment:
Let's propagate errors like this upstream and handle them in the event loop where `apply` is called (i.e. try to avoid panicking deeply inside execution logic).

Reply:
I think the idea here is that this will never panic unless there's developer error, should I still update it?

Reply:
We should make a note a the top of `app` that states this. Looking at the penumbra code, it seems to me that the main motivation to wrap all of this inside `Arc` is to be found here: https://github.com/penumbra-zone/penumbra/blob/07d93f77520cb1ed777c10077ea73d54f532a3fa/crates/core/app/src/app/mod.rs#L219-L233

In other words: they use `Arc` to do (potentially threaded) computations very deeply into mutating state, right before committing it. Except for `check_stateless` and `check_stateful` all other execution is strictly sequential (not even concurrent but really sequential) to the point of it crashing the program because otherwise the invariant is violated.

I have not found evidence to the contrary, but I might have missed it.

Reply:
will add a note. `check_stateless` is for a transaction, it doesn't actually deal with the `Arc<StateDelta>` at all. but yeah when the tx is executed or other state changes are made, they all appear to be sequentlal